File: app/src/main/java/com/dpis/module/ModuleRuntimeReloadAdvisor.java
package com.dpis.module;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

final class ModuleRuntimeReloadAdvisor {
    private static final String PREFS_NAME = "dpis.module_runtime";
    private static final String KEY_ACKED_UPDATE_TIME = "acked_update_time";
    private static final long RUNTIME_LOAD_TOLERANCE_MS = 2_000L;

    private ModuleRuntimeReloadAdvisor() {
    }

    static boolean shouldShowReloadAdvice(Context context) {
        long lastUpdateTime = getLastUpdateTime(context);
        long systemServerLoadedAt = ModuleRuntimeStateReporter.getSystemServerLoadedAt();
        if (!isSystemServerRuntimeOlderThanInstall(lastUpdateTime, systemServerLoadedAt)) {
            return false;
        }
        SharedPreferences preferences = getPreferences(context);
        return preferences.getLong(KEY_ACKED_UPDATE_TIME, 0L) != lastUpdateTime;
    }

    static void markReloadAdviceAcknowledged(Context context) {
        long lastUpdateTime = getLastUpdateTime(context);
        if (lastUpdateTime <= 0L) {
            return;
        }
        getPreferences(context).edit()
                .putLong(KEY_ACKED_UPDATE_TIME, lastUpdateTime)
                .apply();
    }

    static boolean isSystemServerRuntimeOlderThanInstall(long lastUpdateTime,
                                                         long systemServerLoadedAt) {
        if (lastUpdateTime <= 0L || systemServerLoadedAt <= 0L) {
            return false;
        }
        return lastUpdateTime > systemServerLoadedAt + RUNTIME_LOAD_TOLERANCE_MS;
    }

    private static long getLastUpdateTime(Context context) {
        if (context == null) {
            return 0L;
        }
        try {
            PackageInfo packageInfo = context.getPackageManager()
                    .getPackageInfo(context.getPackageName(), 0);
            return packageInfo.lastUpdateTime;
        } catch (PackageManager.NameNotFoundException | RuntimeException ignored) {
            return 0L;
        }
    }

    private static SharedPreferences getPreferences(Context context) {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }
}

File: app/src/main/java/com/dpis/module/runtime/ModuleRuntimeReloadAdvisor.kt
package com.dpis.module.runtime

import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager

object ModuleRuntimeReloadAdvisor {
    private const val PREFS_NAME = "dpis.module_runtime"
    private const val KEY_SHOWN_UPDATE_TIME = "shown_update_time"
    private const val RUNTIME_LOAD_TOLERANCE_MS = 2_000L

    @JvmStatic
    fun shouldShowReloadAdvice(context: Context?): Boolean {
        val lastUpdateTime = getLastUpdateTime(context)
        val systemServerLoadedAt = ModuleRuntimeStateReporter.getSystemServerLoadedAt()
        if (!isSystemServerRuntimeOlderThanInstall(lastUpdateTime, systemServerLoadedAt)) {
            return false
        }
        return getPreferences(context).getLong(KEY_SHOWN_UPDATE_TIME, 0L) != lastUpdateTime
    }

    @JvmStatic
    fun markReloadAdviceShown(context: Context?): Boolean {
        val lastUpdateTime = getLastUpdateTime(context)
        if (lastUpdateTime <= 0L) {
            return false
        }
        // The notice can be followed immediately by a configuration recreation.
        // Commit so the next Activity instance observes that it was already shown.
        return getPreferences(context)
            .edit()
            .putLong(KEY_SHOWN_UPDATE_TIME, lastUpdateTime)
            .commit()
    }

    @JvmStatic
    fun isSystemServerRuntimeOlderThanInstall(
        lastUpdateTime: Long,
        systemServerLoadedAt: Long,
    ): Boolean = lastUpdateTime > 0L &&
        systemServerLoadedAt > 0L &&
        lastUpdateTime > systemServerLoadedAt + RUNTIME_LOAD_TOLERANCE_MS

    private fun getLastUpdateTime(context: Context?): Long {
        val appContext = context ?: return 0L
        return try {
            appContext.packageManager
                .getPackageInfo(appContext.packageName, 0)
                .lastUpdateTime
        } catch (_: PackageManager.NameNotFoundException) {
            0L
        } catch (_: RuntimeException) {
            0L
        }
    }

    private fun getPreferences(context: Context?): SharedPreferences =
        context?.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            ?: error("A context is required after resolving the install timestamp")
}
